README.md
# SubmarineFree 
Free plugins for VCV Rack

![](https://img.shields.io/github/workflow/status/david-c14/SubmarineFree/Release?label=Release) ![](https://img.shields.io/github/workflow/status/david-c14/SubmarineFree/Develop?label=Develop) ![](https://img.shields.io/github/v/release/david-c14/SubmarineFree?label=Latest) ![](https://img.shields.io/github/release-date/david-c14/SubmarineFree?label=Released) ![](https://img.shields.io/badge/License-GPLv3-blue.svg)

[.plan (what I'm working on)](https://github.com/david-c14/SubmarineFree/issues/23)

# [Manual](https://github.com/david-c14/SubmarineFree/blob/main/manual/index.md)

# [Releases](https://github.com/david-c14/SubmarineFree/releases)

# [License](gpl-3.0.txt)
Source code licensed under GPL-3.0 by carbon14 (David O'Rourke)

Some portions of this source code are based on code © 2016 by Andrew Belt


changelog.md
# Changelog

## v2.0.4
- 04-Jan-2022 Update CI pipelines to SDK v2.0.5
- 04-Jan-2022 Fix resizing issue with TD-316
- 04-Jan-2022 Fix bug #193 incorrect wiring for AG-202, OG-202 and XG-202 devices
- 22-Dec-2021 Add AO-101 module
- 20-Dec-2021 Fix bug #186 missing algorithms is AO-1xx manual
- 18-Dec-2021 Fix bug #185 html tables in manual

## v2.0.3
- 13-Dec-2021 Fix unitialized variable in XF-301

## v2.0.2
- 06-Dec-2021 Fix bug #175 High CPU usage in SS-212
- 04-Dec-2021 Minor fixes in manual
- 04-Dec-2021 Fix bug #172 Visuals not updating on TD-410 and TD-510 in a timely manner

## v2.0.1
- 30-Nov-2021 Add LA-216 Polyphonic Logic Analyser

## v2.0.0
- 11-Oct-2021 Port shared components to v2
- 11-Oct-2021 Port Blanking plates to v2
- 14-Oct-2021 Port LA-108 to v2
- 15-Oct-2021 Port AG-xxx OG-xxx NG-xxx XG-xxx LD-xxx and PG-xxx to v2
- 15-Oct-2021 Persist DS suite voltage range between devices. When voltage is changed, new devices use the same option
- 15-Oct-2021 Port BB-xxx DN-xxx FF-xxx OA-xxx to v2
- 25-Oct-2021 Port Text devices and Wiremanager
- 26-Oct-2021 Port HS-101 and EO-102 to v2
- 26-Oct-2021 Port EN-104 PO-101 PO-102 PO-202 and SN-101 to v2
- 28-Oct-2021 Port WK-101 and WK-205 to v2  Deprecated Torpedo in WK devices and replace with extension modules
- 28-Oct-2021 Add Module Link lights for expander modules
- 29-Oct-2021 Port LT-116 to v2
- 29-Oct-2021 Port XF-xxx to v2
- 29-Oct-2021 Port VU meters to v2, including new backlights
- 03-Nov-2021 Make DO-1xx devices more usable, improve connection labelling, add shuffle and copy/paste
- 05-Nov-2021 Add visuals to TD-xxx, DO-1xx and WM-102 devices in the module browser
- 05-Nov-2021 Add Device copy/paste and undo to AO-1xx devices
- 12-Nov-2021 Add expansion capabilities to VM-1xx and VM-2xx devices
- 14-Nov-2021 Use keywords to make library searching more effective
- 14-Nov-2021 Add VM-104 Quad digital VU-meter
- 19-Nov-2021 Add VM-204 Quad vintage VU-meter
- 22-Nov-2021 Fix rendering bug in VM-1xx devices

## v1.1.10
- 10-Dec-2020 Update images in manual
- 26-Nov-2020 Add exponential response to VM-needle
- 26-Nov-2020 Fix bug rendering WM-102 in module browser

## v1.1.9
- 23-Nov-2020 Add DS-2 series polyphonic logic gates
- 22-Nov-2020 Add OA-1xx Operation Amplifiers
- 22-Nov-2020 Add visuals menu to scopes (missing since addition of "Match Cable Colors")
- 20-Nov-2020 VM-1xx and VM-2xx VU Meters

## v1.1.8
- 12-Sep-2020 Bug fix #145 EN-104 inputs

## v1.1.7
- 09-Sep-2020 Bug fix #141 TD-3/4/5 minimum width
- 05-Sep-2020 Add EN-104 Envelope generator
- 30-Jul-2020 Extend time-base of HS-101 down to as short as 1/1000 of a second
- 30-Jul-2020 Extend time-base of LA-108 to as long as 128 seconds
- 24-Jul-2020 Bug fix for MZ-909 manual. Thank you Ryan Monro.

## v1.1.6
- 19-Jul-2020 Added MZ-909 Masterizer
- 18-Jul-2020 Correct urls that refer to master
- 15-Jul-2020 Make devices work with Modular Fungi LightsOff
- 04-Jul-2020 Bug fix - #127 BP-102 incorrect behaviour
- 16-May-2020 WM-102 persist data in Module not in Widget
- 12-May-2020 Deal with warning over hidden send methods in torpedo
- 09-May-2020 Correct use of integer abs to float fabs
- 09-May-2020 Move from deprecated SVGWidget to SvgWidget

## v1.1.5
- 07-May-2020 Issue #116 - Sum Polyphonic inputs for AO-XXX and XF-XXX
- 30-Apr-2020 TD-510 Vertical multiple label textbox
- 30-Apr-2020 TD-410 Add font size option
- 29-Apr-2020 TD-410 Add duplicate label option
- 29-Apr-2020 TD-410 persist data in Module not in Widget
- 25-Apr-2020 TD-202 persist data in Module not in Widget
- 25-Apr-2020 TD-116 persist data in Module not in Widget
- 24-Apr-2020 TD-316 persist data in Module not in Widget
- 23-Apr-2020 TD-410 Make resizeable
- 23-Apr-2020 TD-316 Make resizeable
- 15-Apr-2020 WM-101 persist width in Module not in Widget
- 14-Apr-2020 HS-101 Trigger output pulse width changed to 1ms

## v1.1.4
- 06-Apr-2020 Tidy plugin manifest to suit library
- 02-Apr-2020 Add HS-101 Hi-Res Storage Scope
- 01-Apr-2020 Add TD-410 multiple label textbox
- 01-Apr-2020 Add bias compensator to XF-XXX cross faders
- 01-Apr-2020 Tidy up filenames
- 01-Apr-2020 Tidy up source code for XF modules
- 01-Apr-2020 Add XF-302 Small Form Factor Stereo Cross Fader
- 31-Mar-2020 Make WM-102 billboard automatically track WM-101

## v1.1.3
- 30-Mar-2020 Add WM-102 Wire Manager Billboard
- 30-Mar-2020 Fix memory leak and crash in WM-101 color labels

## v1.1.2
- 25-Mar-2020 Jim Allman added names to colors - and billboard view
- 21-Mar-2020 Jim Allman added some popular color conventions as collections
- 21-Mar-2020 Add undo to load collections
- 21-Mar-2020 Bug fix - Undo on move collection down was not working
- 18-Mar-2020 Issue #80 - Save and load collections in WM-101
- 17-Mar-2020 Add LT-116 Linear Transformations
- 17-Mar-2020 Bug fix #78 - Color one wire uses stale color
- 28-Jan-2020 Replace Digital Ports (Blue) with Analog (Silver) on EO-102
- 21-Jan-2020 Issue #50 - Protect internal classes with anonymous namespaces

## v1.1.1
- 02-Jan-2020 Secret Santa 2019 is now obsolete
- 02-Jan-2020 Make WK-XXX case insensitive when searching for Scala files
- 24-Dec-2019 Add Smooth Noise SN-101 module
- 19-Dec-2019 Two typos in the manual kindly corrected by Darío Hereñú
- 17-Dec-2019 Improve Manual for DO-1xx devices
- 07-Dec-2019 Make DO-1xx connectors scrolling consistent between connectors and across devices
- 07-Dec-2019 Bug fix #54 - DN-112 does not produce maximal sequence
- 03-Dec-2019 Remove unnecessary base constructor calls

## v1.1.0
- 26-Nov-2019 Add Secret Santa 2019
- 26-Nov-2019 Add DO-1xx Digital Operator Devices
- 17-Nov-2019 Fix module id on LA-108 front panel
- 27-Oct-2019 Fix typo in manual (courtesy of kant)
- 22-Oct-2019 Bug fix #48 - Conflicting class names
- 19-Oct-2019 PO-xxx devices moved to SIMD code
- 19-Oct-2019 Relicenced to GPL-3.0
- 18-Oct-2019 Bug fix - Color 1 wire not working responding to existing cables

src/SS.cpp
#include "SubmarineFree.hpp"

struct SS_112 : Module {
	static constexpr int deviceCount = 12;
	SS_112() : Module(0, deviceCount, 0, 0) {}
};

struct SS112 : ModuleWidget {
	SS112(SS_112 *module) : ModuleWidget(module) {
		setPanel(SVG::load(assetPlugin(plugin, "res/SS-112.svg")));
		for (int i = 0; i < SS_112::deviceCount; i++) {
			addInput(Port::create<sub_port>(Vec(2.5,19 + i * 29), Port::INPUT, module, i));
		}
	}
};

struct SS_208 : Module {
	static constexpr int deviceCount = 8;
	SS_208() : Module(0, 0, deviceCount, 0) {
		outputs[0].value = M_PI;
		outputs[1].value = 2 * M_PI;
		outputs[2].value = M_E;
		outputs[3].value = M_SQRT1_2;
		outputs[4].value = M_SQRT2;
		outputs[5].value = powf(3.0f, 0.5f);
		outputs[6].value = powf(5.0f, 0.5f);
		outputs[7].value = powf(7.0f, 0.5f);
	}
};

struct SS208 : ModuleWidget {
	SS208(SS_208 *module) : ModuleWidget(module) {
		setPanel(SVG::load(assetPlugin(plugin, "res/SS-208.svg")));
		for (int i = 0; i < SS_208::deviceCount; i++) {
			addOutput(Port::create<sub_port>(Vec(2.5,19 + 43 * i), Port::OUTPUT, module, i));
		}
	}
};

struct SS_212 : Module {
	static constexpr int deviceCount = 12;
	int v = 0;
	void setValues() {
		for (int i = 0; i < deviceCount; i++) {
			outputs[i].value = v + 1.0f * i / 12.0f;
		}
	}

	SS_212() : Module(0, 0, deviceCount, 0) {
		setValues();
	}

	json_t *toJson() override {
		json_t *rootJ = json_object();
		json_object_set_new(rootJ, "octave", json_integer(v));
		return rootJ;
	}

	void fromJson(json_t *rootJ) override {
		json_t *intJ = json_object_get(rootJ, "octave");
		if (intJ)
			v = json_integer_value(intJ);
		setValues();
	}
};

struct SS212 : ModuleWidget {
	SS212(SS_212 *module) : ModuleWidget(module) {
		setPanel(SVG::load(assetPlugin(plugin, "res/SS-212.svg")));
		for (int i = 0; i < SS_212::deviceCount; i++) {
			addOutput(Port::create<sub_port>(Vec(2.5,19 + i * 29), Port::OUTPUT, module, i));
		}
	}

	void appendContextMenu(Menu *menu) override;
};

struct SSMenuItem : MenuItem {
	SS_212 *ss_212;
	int v;
	void onAction(EventAction &e) override {
		ss_212->v = v;
		ss_212->setValues();
	}
	void step() override {
		rightText = CHECKMARK(ss_212->v == v);
	}
};

void SS212::appendContextMenu(Menu *menu) {
	char label[20];
	menu->addChild(MenuEntry::create());
	SS_212 *ss_212 = dynamic_cast<SS_212*>(this->module);
        assert(ss_212);
	for (int i = -5; i < 5; i++) {
		sprintf(label, "Octave %d", i);
		SSMenuItem *menuItem = MenuItem::create<SSMenuItem>(label);
		menuItem->ss_212 = ss_212;
		menuItem->v = i;
		menu->addChild(menuItem);
	}
}

struct SS_221 : Module {
	static constexpr int deviceCount = 21;
	SS_221() : Module(0, 0, deviceCount, 0) {
		for (int i = 0; i < deviceCount; i++) {
			outputs[i].value = 10.0f - i;
		}
	}
};

struct SS221 : ModuleWidget {
	SS221(SS_221 *module) : ModuleWidget(module) {
		setPanel(SVG::load(assetPlugin(plugin, "res/SS-221.svg")));
		for (int i = 0; i < SS_221::deviceCount; i++) {
			addOutput(Port::create<sub_port>(Vec(2.5 + 45 * (i % 2),19 + i * 16), Port::OUTPUT, module, i));
		}
	}
};

struct SS_220 : Module {
	static constexpr int deviceCount = 12;
	static constexpr int deviceSetCount = 10;
	SS_220() : Module(0, 0, deviceCount * deviceSetCount, 0) {
		for (int j = 0; j < deviceSetCount; j++) {
			for (int i = 0; i < deviceCount; i++) {
				outputs[j * deviceCount + i].value = (j - 5.0f) + 1.0f * i / 12.0f;
			}
		}
	}
};

struct SS220 : ModuleWidget {
	SS220(SS_220 *module) : ModuleWidget(module) {
		setPanel(SVG::load(assetPlugin(plugin, "res/SS-220.svg")));
		for (int j = 0; j < SS_220::deviceSetCount; j++) {
			for (int i = 0; i < SS_220::deviceCount; i++) {
				addOutput(Port::create<sub_port>(Vec(2.5 + 30 * j, 19 + i * 29), Port::OUTPUT, module, j * SS_220::deviceCount + i));
			}
		}
	}
};

Model *modelSS112 = Model::create<SS_112, SS112>("SubmarineFree", "SS-112", "SS-112 12 Input Sinks", UTILITY_TAG);
Model *modelSS208 = Model::create<SS_208, SS208>("SubmarineFree", "SS-208", "SS-208 8 Irrational Output Voltage Sources", UTILITY_TAG);
Model *modelSS212 = Model::create<SS_212, SS212>("SubmarineFree", "SS-212", "SS-212 12 Chromatic Output Voltage Sources", UTILITY_TAG);
Model *modelSS220 = Model::create<SS_220, SS220>("SubmarineFree", "SS-220", "SS-220 120 Chromatic Output Voltage Sources", UTILITY_TAG);
Model *modelSS221 = Model::create<SS_221, SS221>("SubmarineFree", "SS-221", "SS-221 21 Output Voltage Sources", UTILITY_TAG);
